core/storage/question/gae_models.py

```diff
@@ -389,7 +390,7 @@ def get_offset(query):
                     new_question_skill_link_models.remove(model)
 
             if len(new_question_skill_link_models) >= question_count_per_skill:
-                new_question_skill_link_models = utils.get_random_sample(
+                new_question_skill_link_models = utils.random.sample(
```

Comment:
This looks completely wrong: utils doesn't have a member named random. But could you please investigate why the backend tests are not catching it, and appear to be passing despite this?

Reply:
I have tested this and it indeed works. What this does is import the random that is imported by the utils file. This is how I set the seed in the gae_models_test file. Since you asked me to use random.sample I thought it would be better if I use the one from utils file. I will change it if this method is wrong.

Reply:
Oh, I see, don't do that please. Instead, in the test file, swap out the call to random.sample with your own implementation.

Reply:
(Also, thanks for explaining!)

Reply:
Just to be clear you want me to implement random.sample (like I did previously in the utils file). Or you want me to directly import random in gae_models and use random.sample?

Reply:
Neither. In the tests, use self.swap() to swap random.sample() with a mock method that does something deterministic.

Reply:
If we use random.sample directly I don't think it is possible to swap the method with the mock method.
i.e self.swap(random,'sample', mock_random_sample) doesn't work

Reply:
It should work. Why doesn't it work?

Reply:
I am not sure maybe because it is a python inbuilt library and not one which we implemented. But instead of using swap we can set the seed for random right, It should work fine.

Reply:
@rohitkatlaa -- when you say "X doesn't work", please always, at the very least, provide the code you tried, and the stack trace, and what you did to try and figure out where the error in the stack trace comes from. Otherwise, I won't be able to help you, because I don't know what you have actually tried to do. (I can't tell at the moment whether the "doesn't work" is due to an actual issue or an error in implementation.)

Reply:
I tied the following
```
def mock_random_sample(alist, num):
   if num >= len(alist):
        return alist
        alist.sort(key= question_id)
    return [1,2,3]

self.swap(random, 'sample', mock_random_sample)```
Now the random.sample function is supposed to return [1,2,3] but instead it executed in the default way. I have also checked that the random.sample is being reached in the code. Thus I came to a conclusion that maybe inbuilt functions can't be swapped. A possible workaround for this can be to implement our sample function in the utils and swap it in the test.